### PR TITLE
Allow arbitrary objects in state_dicts

### LINKED ISSUE #62094: [FR] Support for non-Parameter, non-Buffer objects in state_dict
## 🚀 Feature
Support adding arbitrary things to a module's `state_dict`. Currently, serialization of a module is tied to what is stored in its `state_dict`, and there is no public API for storing arbitrary things in the `state_dict`. This limits control over what is serialized.

### Use Cases
1. Serialize arbitrary objects within a module (flags, integers, etc.)
    * See #51471
2. Define "buffers" that are serialized but don't take part in device moves / floating pointing dtype conversions (e.g. `to()`, `cuda()`, etc.)
    * No current way to do this; serialization can be turned off for buffers by marking them as non-persistent, but they cannot be left out of device moves / floating point dtype conversions
    * Equivalent to allowing non-Parameter, non-buffer Tensors in `state_dict`
3. Support custom Tensor types within modules (e.g. `DiagTensor`, `ShardedTensor`, etc.)
    * Part of the solution is to allow custom Tensor types to be serialized
    * Full solution would allow custom Tensor types to act as Parameters for use in optimization (likely a separate discussion)

### Possible solutions
One possible solution is to add `state_dict` hooks to the public API. Currently we have `_register_state_dict_hook()` / `_register_load_state_dict_pre_hook()`, which are limited since they do not receive the module instance itself. More direct mechanisms should be considered as well.

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #62094

Introduces functionality for adding arbitrary objects to module state_dicts. To take advantage of this, the following functions can be defined on a module:
* `get_extra_state(self) -> dict` - Returns a dict defining any extra state this module wants to save
* `set_extra_state(self, state)` - Subsumes the given state within the module

In the details, a sub-dictionary is stored in the state_dict under the key `_extra_state` for each module that requires extra state.

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -531,6 +533,35 @@ def get_buffer(self, target: str) -> "Tensor":
...
 
+    def get_extra_state(self) -> Dict[str, Any]:
+        """
+        Returns any extra state to include in the module's state_dict.
+        Implement this and a corresponding :func:`set_extra_state` for your module
+        if you need to store extra state. This function is called when building the
+        module's `state_dict()`.
+
+        Note that extra state should be pickleable to ensure working serialization
+        of the module. We only provide provide backwards compatibility guarantees
```
Comment (albanD) [inline torch/nn/modules/module.py:544]: nit: of the module -> of the state_dict

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -531,6 +533,35 @@ def get_buffer(self, target: str) -> "Tensor":
...
+        for serializing Tensors; other objects may break backwards compatibility if
+        their serialized pickle form changes.
```
Comment (albanD) [inline torch/nn/modules/module.py:546]: nit pickled

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -531,6 +533,35 @@ def get_buffer(self, target: str) -> "Tensor":
...
+
+        Returns:
+            dict or OrderedDict: Any extra state to store in the module's state_dict
```
Comment (albanD) [inline torch/nn/modules/module.py:549]: Aren't all dicts OrderedDicts now anyways?

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -531,6 +533,35 @@ def get_buffer(self, target: str) -> "Tensor":
...
+        """
+        raise NotImplementedError()
```
Comment (albanD) [inline torch/nn/modules/module.py:551]: nit: we usually have a message in the not implemented errors explaining why you're hitting this and which function you should implement.
If this is supposed to never be called, we can make the error a RuntimeError and ask the user to open an issue with a repro if they actually hit that error.

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -531,6 +533,35 @@ def get_buffer(self, target: str) -> "Tensor":
 
         return buffer
 
+    def get_extra_state(self) -> Dict[str, Any]:
```
Comment (albanD) [inline torch/nn/modules/module.py:536]: Does this really have to return a Dict? Why not Any?

File: torch/nn/modules/module.py [hunk 6]
```diff
@@ -1227,6 +1258,11 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
         for name, buf in self._buffers.items():
             if buf is not None and name not in self._non_persistent_buffers_set:
                 destination[prefix + name] = buf if keep_vars else buf.detach()
+        extra_state_key = prefix + _EXTRA_STATE_KEY_SUFFIX
+        if getattr(self.__class__, "get_extra_state", Module.get_extra_state) is not Module.get_extra_state:
+            destination[extra_state_key] = OrderedDict()
+            for name, obj in self.get_extra_state().items():
+                destination[extra_state_key][name] = obj
```
Comment (albanD) [inline torch/nn/modules/module.py:1265]: Why do you write these items by hand instead of just doing `destination[extra_state_key] = self.get_extra_state()` ?

File: torch/nn/modules/module.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/module.py:549]: True; changed to just read 'dict'

Comment (jbschlosser) [reply torch/nn/modules/module.py:549]: Changed again to read 'object'

File: torch/nn/modules/module.py (hunk 5 above)
Comment (jbschlosser) [reply torch/nn/modules/module.py:536]: Changed to Any

File: torch/nn/modules/module.py (hunk 6 above)
Comment (jbschlosser) [reply torch/nn/modules/module.py:1265]: Fixed

**Review by albanD — APPROVED** [review]
Looks good!
